ern/graphql.go

```diff
@@ -0,0 +1,553 @@
+// This file is part of the go-meta library.
+//
+// Copyright (C) 2017 JAAK MUSIC LTD
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// If you have any questions please contact yo@jaak.io
+
+package ern
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-ipld-format"
+	"github.com/meta-network/go-meta"
+)
+
+// Resolver is a general purpose GraphQL resolver function,
+// retrieves data from a META store and SQLite3 index
+type Resolver struct {
+	db    *sql.DB
+	store *meta.Store
+}
+
+// NewResolver returns a new resolver for returning data from the given
+// META store and SQLite3 index
+func NewResolver(db *sql.DB, store *meta.Store) *Resolver {
+	return &Resolver{db, store}
+}
+
+// GraphQLSchema schema definition for PartyDetails on the ERN index
+const GraphQLSchema = `
+	schema {
+		query: Query
+	}
+
+	type Query {
+		partyDetails(id: String, name: String): [PartyDetails]!
+		soundRecording(id: String, title: String): [SoundRecording]!
+		release(id: String, title: String): [Release]!
+	}
+
+	type PartyDetails {
+		cid: String!
+		partyID: String!
+		fullName: String!
+	}
+
+	type ResourceContributor {
+		partyDetails: PartyDetails!
+		role: String!
+	}
+
+	type SoundRecording {
+		artistName: String!
+		# contributors: [ResourceContributor]!
+		genre: String!
+		# parentalWarningType: String
+		resourceReference: String!
+		# subGenre: String
+		soundRecordingId: String!
+		territoryCode: String!
+		title: String!
+	}
+
+	type Release {
+		artistName: String!
+		displayTitle: String!
+		releaseId: String
+		genre: String
+		releaseType: String
+	}
+`
+
+// partyDetailsArgs query arguments for PartyDetails query
+type partyDetailsArgs struct {
+	Name *string
+	ID   *string
+}
+
+// DecodeResolverObj finds a value within a MetaObject and decode it
+// to a specific schema, then return result
+// metaObj - the meta object to be decoded
+// v - the schema defintion to decode to
+// path - the tree path (that is found in the source data document (ERN))
+func DecodeResolverObj(i *Resolver, metaObj *meta.Object, v interface{}, path ...string) (err error) {
```

Comment:
Pass in `*meta.Store` rather than `*Resolver` then this can be consolidated with the `DecodeIndexObj` function.

Reply:
Thanks